### PR TITLE
implement interfaces and unions 

### PR DESCRIPTION
**Pull request by a-alle**
# Description

This PR contains:
- distinction between Interfaces and Unions as CompositeEntities (CompositeEntity now an interface implemented by concrete classes InterfaceEntity and UnionEntity)
- InterfaceEntity and UnionEntity added to the SchemaModel generation logic
- InterfaceEntity contain attributes and annotations (but no relationships)
- relationships can have InterfaceEntity as source
- no annotation inheritance from Interface to Concrete Entity
- tests updates

Caveats:
- interface fields cannot have annotations (except for some annotations on relationshipProperties interfaces)
- interfaces have concrete entities but they can be implemented by other interfaces - check if this needs to be modelled
 
## Complexity

Complexity: Low

# Issue

> **Note**
>
>  Please link to the GitHub issue(s) in which the proposal for this work was discussed
>  
>  To link to multiple issues, use full syntax for each, for example `Closes #1, closes #2, closes #3`

Closes 

# Checklist

The following requirements should have been met (depending on the changes in the branch):

- [ ] Documentation has been updated
- [ ] TCK tests have been updated
- [ ] Integration tests have been updated
- [ ] Example applications have been updated
- [ ] New files have copyright header
- [ ] CLA (https://neo4j.com/developer/cla/) has been signed

### CODE REVIEW

**Comment by neo4j-team-graphql** [comment]
Made some changes to the Neo4j GraphQL Toolbox? Or made some changes to the core library perhaps?

Check out your changes at https://neo4j-graphql-toolbox-3812.surge.sh

Use for instance a local dbms hosted on [Neo4j Desktop](https://neo4j.com/download/) or a cloud-hosted [Aura Free](https://neo4j.com/cloud/aura-free/) dbms.

**Comment by neo4j-team-graphql** [comment]
# Performance Report

No Performance Changes


<details>
    <summary>Show Full Table</summary>
    
| name | dbHits | old dbHits | time (ms) | old time (ms) | maxRows |
| ---- | ------ | ---------- | --------- | ------------- | ------- |
|  aggregations.TopLevelAggregate | 3403 | 3403 | 42 | 45 | 1134 |
|  aggregations.NestedAggregation | 16553 | 16553 | 85 | 94 | 2174 |
|  aggregations.AggregationWithWhere | 11979 | 11979 | 53 | 58 | 2174 |
|  aggregations.AggregationWhereWithinNestedRelationships | 22116987 | 22116987 | 3509 | 3770 | 2008534 |
|  aggregations.AggregationWhereWithinNestedConnections | 22116987 | 22116987 | 3200 | 3280 | 2008534 |
|  aggregations.NestedCountFromMovieToActors | 9734 | 9734 | 46 | 82 | 2174 |
|  aggregations.NestedCountFromActorsToMovie | 9937 | 9937 | 54 | 87 | 2174 |
|  aggregations.DeeplyNestedCount | 13070331 | 13070331 | 5225 | 5117 | 2008534 |
|  batch-create.BatchCreate | 4200 | 4200 | 202 | 285 | 600 |
|  batch-create.BatchCreateSmall | 77 | 77 | 141 | 153 | 11 |
|  connect.createAndConnect | 14424 | 14424 | 225 | 258 | 3003 |
|  connections.Connection | 14082 | 14082 | 93 | 77 | 2174 |
|  connections.NestedConnection | 45470 | 45472 | 196 | 166 | 4516 |
|  connections.ConnectionWithSort | 3284 | 3284 | 85 | 83 | 1040 |
|  connections.ConnectionWithSortAndCypher | 16325 | 16325 | 117 | 145 | 2174 |
|  create.SimpleMutation | 7 | 7 | 67 | 75 | 1 |
|  cypher-directive.TopLevelMutationDirective | 1135 | 1135 | 34 | 38 | 1134 |
|  delete.SimpleDelete | 19401 | 19401 | 304 | 252 | 1040 |
|  delete.NestedDeleteInUpdate | 19198 | 19198 | 224 | 191 | 1183 |
|  2925.SingleRelationshipFilter | 6468 | 6468 | 62 | 62 | 1040 |
|  2925.NestedSingleRelationshipFilter | 22900 | 22900 | 92 | 143 | 2174 |
|  2925.SingleRelationshipRequiredFilter | 5201 | 5201 | 44 | 54 | 1040 |
|  2925.NestedSingleRelationshipRequiredFilter | 9361 | 9361 | 68 | 83 | 1040 |
|  query.SimpleQuery | 3121 | 3121 | 37 | 25 | 1040 |
|  query.SimpleQueryWithRelationship | 16162 | 16162 | 49 | 51 | 2174 |
|  query.QueryWhere | 9710 | 9697 | 41 | 47 | 2165 |
|  query.SimpleQueryWithNestedWhere | 9888 | 9875 | 59 | 66 | 2165 |
|  query.Nested | 10096041 | 10096041 | 10657 | 11567 | 2008534 |
|  query.NestedWithFilter | 10074401 | 10074401 | 10809 | 11038 | 2004000 |
|  query.OrFilterOnRelationships | 41999 | 40848 | 198 | 216 | 1641 |
|  query.OrFilterOnRelationshipsAndNested | 35910 | 35426 | 233 | 333 | 1641 |
|  query.QueryWithNestedIn | 14581 | 13421 | 77 | 90 | 1611 |
|  query.NestedConnectionWhere | 9834 | 9834 | 81 | 119 | 2174 |
|  query.DeeplyNestedConnectionWhere | 10044 | 9896 | 97 | 136 | 2174 |
|  query.DeeplyNestedWithRelationshipFilters | 19248 | 19116 | 185 | 188 | 1630 |
|  query.NestedWithRelationshipSingleFilters | 3881 | 3881 | 251 | 252 | 1134 |
|  query.Fulltext | 80 | 80 | 38 | 52 | 16 |
|  query.FulltextWithNestedQuery | 587 | 587 | 50 | 59 | 84 |
|  sorting.SortMultipleTypes | 3515 | 3515 | 82 | 104 | 1040 |
|  sorting.SortMultipleTypesWithCypherWithCypher | 14526 | 14491 | 109 | 150 | 2174 |
|  sorting.SortOnNestedFields | 14082 | 14082 | 56 | 83 | 2174 |
|  sorting.SortDeeplyNestedFields | 43198 | 43198 | 105 | 132 | 4516 |
|  sorting.SortWithTopLevelCypher | 3396 | 3396 | 68 | 75 | 1040 |
|  unions.SimpleUnionQuery | 321 | 321 | 69 | 74 | 35 |
|  unions.SimpleUnionQueryWithMissingFields | 293 | 293 | 67 | 74 | 35 |
|  unions.NestedUnion | 410637 | 410637 | 298 | 389 | 33033 |
|  unions.NestedUnionWithMissingFields | 384611 | 384611 | 307 | 445 | 33033 |
|  update.NestedUpdate | 16143 | 16143 | 171 | 133 | 2002 |

</details>
        
Old Schema Generation: 48.225s
Schema Generation: 49.063s
Old Subgraph Schema Generation: 1:01.514 (m:ss.mmm)
Subgraph Schema Generation: 58.981s

File: packages/graphql/src/schema-model/Neo4jGraphQLSchemaModel.ts
Comment (angrykoala) [inline packages/graphql/src/schema-model/Neo4jGraphQLSchemaModel.ts:91]: I thought the idea of having a composite entity is to avoid having both :thinking:

File: packages/graphql/src/schema-model/entity/ConcreteEntity.ts
Comment (angrykoala) [inline packages/graphql/src/schema-model/entity/ConcreteEntity.ts:83]: I guess that this is due to a workflow a bit like:

* Add all attributes
* Remove those that are relationship

Is there a way of doing these checks outside the schema model? Ideally, all of this logic exists in the factories, so the schema model is somehow static

File: packages/graphql/src/translate/projection/subquery/translate-cypher-directive-projection.ts
Comment (angrykoala) [inline packages/graphql/src/translate/projection/subquery/translate-cypher-directive-projection.ts:95]: I'm thinking, how about holding this information in the entity class itself?

```
entity.isComposite()
```

To be clear, I think we need to rethink a bit the whole composite entity, but maybe it will be a bit cleaner using a method in the entity itself, so we don't need the whole schemaModel for this check

File: packages/graphql/src/schema-model/relationship/model-adapters/RelationshipAdapter.ts
Comment (angrykoala) [inline packages/graphql/src/schema-model/relationship/model-adapters/RelationshipAdapter.ts:93]: This logic is a bit ugly, I think it may be pointing to a bit of a design issue with how the adapters work tbh :sweat_smile:

File: packages/graphql/src/schema-model/generate-model.test.ts
Comment (angrykoala) [inline packages/graphql/src/schema-model/generate-model.test.ts:552]: :eyes:

**Review by angrykoala — APPROVED** [review]
Couple of comments, Looks good to me, although I'm not sure what the reasoning behind the move towards Union + Interface instead of Composite

File: packages/graphql/src/schema-model/Neo4jGraphQLSchemaModel.ts
Comment (a-alle) [reply packages/graphql/src/schema-model/Neo4jGraphQLSchemaModel.ts:91]: The problem is they are very different and behave differently. All they have in common is a name and conceptually "dependant" concrete entities, however even these are very different: a union can only be made up of concrete entities, while an interface can also be implemented by interface entities (not unions).

File: packages/graphql/src/schema-model/entity/ConcreteEntity.ts
Comment (a-alle) [reply packages/graphql/src/schema-model/entity/ConcreteEntity.ts:83]: so you mean exposing deleteAttribute and doing this reassignment in the generate-model?

File: packages/graphql/src/translate/projection/subquery/translate-cypher-directive-projection.ts
Comment (a-alle) [reply packages/graphql/src/translate/projection/subquery/translate-cypher-directive-projection.ts:95]: I like this idea

File: packages/graphql/src/schema-model/Neo4jGraphQLSchemaModel.ts
Comment (angrykoala) [reply packages/graphql/src/schema-model/Neo4jGraphQLSchemaModel.ts:91]: mmm. If that's the case, maybe we could scratch composite entities altogether :thinking: Or do you think it these are still useful?

File: packages/graphql/src/schema-model/entity/ConcreteEntity.ts
Comment (angrykoala) [reply packages/graphql/src/schema-model/entity/ConcreteEntity.ts:83]: Mmm, I'm thinking more of figuring out which are relationships and which are attributes before setting them into the schema model.

I'm just saying that because for me it is easier to reason about a stateless model than to modify it after creating it. Just as an alternative to consider :sweat_smile:

Comment (a-alle) [reply packages/graphql/src/schema-model/entity/ConcreteEntity.ts:83]: ah yes I see what you mean - I agree with you that ideally this should be the case. However, that requires changing the way the schema model is being generated, which I plan to do but I wanted to have interfaces and unions in first, in order to unblock the big refactors as quickly as possible. I will assess this!

File: packages/graphql/src/schema-model/Neo4jGraphQLSchemaModel.ts
Comment (a-alle) [reply packages/graphql/src/schema-model/Neo4jGraphQLSchemaModel.ts:91]: yes true.. As of now the CompositeEntity is just a kind of logical grouping, I was keeping it in case it proves useful as it doesn't seem to hurt but I could also remove it and add it back in case we want it at some point. Do you prefer to get rid of it?

File: packages/graphql/src/schema-model/entity/ConcreteEntity.ts
Comment (a-alle) [reply packages/graphql/src/schema-model/entity/ConcreteEntity.ts]: you can revisit this - I changed it from an after-the-fact "hydrate" method that mutates the class to during construction.

File: packages/graphql/src/schema-model/attribute/AttributeType.ts
Comment (a-alle) [inline packages/graphql/src/schema-model/attribute/AttributeType.ts:98]: @MacondoExpress something to consider

**Review by MacondoExpress — APPROVED** [review]
LGTM!

**Comment by mjfwebb** [comment]
Closed in favour of #3889 which now targets dev instead of 4.0.0 branch.

**Pull request closed without merging**
